libfwupd/fwupd-common.c

```diff
@@ -136,34 +140,83 @@ fwupd_checksum_get_best (GPtrArray *checksums)
 	return NULL;
 }
 
-static gchar *
-fwupd_build_user_agent_os_release (void)
+/**
+ * fwupd_build_distro_hash:
+ * @error: A #GError or %NULL
+ *
+ * Loads information from the system os-release file.
+ **/
+static GHashTable *
+fwupd_build_distro_hash (GError **error)
 {
-	const gchar *keys[] = { "NAME=", "VERSION_ID=", "VARIANT=", NULL };
-	const gchar *values[] = { NULL, NULL, NULL, NULL };
-	g_autofree gchar *os_release = NULL;
+	GHashTable *hash;
+	const gchar *filename = NULL;
+	const gchar *paths[] = { "/etc/os-release", "/usr/lib/os-release", NULL };
+	g_autofree gchar *buf = NULL;
 	g_auto(GStrv) lines = NULL;
-	g_autoptr(GPtrArray) ids_os = g_ptr_array_new ();
 
-	/* get raw data then parse each line */
-	if (!g_file_get_contents ("/etc/os-release", &os_release, NULL, NULL)) {
-		if (!g_file_get_contents ("/usr/lib/os-release", &os_release, NULL, NULL))
-			return NULL;
+	/* find the correct file */
+	for (guint i = 0; paths[i] != NULL; i++) {
+		g_debug ("looking for os-release at %s", paths[i]);
+		if (g_file_test (paths[i], G_FILE_TEST_EXISTS)) {
+			filename = paths[i];
+			break;
+		}
 	}
-	lines = g_strsplit (os_release, "\n", -1);
+	if (filename == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_READ,
+				     "No os-release found");
+		return NULL;
+	}
+
+	/* load each line */
+	if (!g_file_get_contents (filename, &buf, NULL, error))
+		return NULL;
+	hash = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+	lines = g_strsplit (buf, "\n", -1);
 	for (guint i = 0; lines[i] != NULL; i++) {
-		for (guint j = 0; keys[j] != NULL; j++) {
-			if (g_str_has_prefix (lines[i], keys[j])) {
-				values[j] = lines[i] + strlen (keys[j]);
-				break;
-			}
+		gsize len, off = 0;
+		g_auto(GStrv) split = NULL;
+
+		/* split up into sections */
+		split = g_strsplit (lines[i], "=", 2);
+		if (g_strv_length (split) < 2)
+			continue;
+
+		/* remove double quotes if set both ends */
+		len = strlen (split[1]);
+		if (len == 0)
+			continue;
+		if (split[1][0] == '\"' && split[1][len-1] == '\"') {
+			off++;
+			len -= 2;
 		}
+		g_hash_table_insert (hash,
+				     g_steal_pointer (&split[0]),
```

Comment:
Hmm. If you steal the pointer here then [`g_strfreev`](https://github.com/GNOME/glib/blob/master/glib/gstrfuncs.c#L2490) will stop freeing here and you leak `split[1]`, won't it?

Reply:
I don't know, I don't know if `g_strfreev` "knows" the size of the array. valgrind seems to not show any leak, but I've changed it to be a `g_strdup` to be less ambiguous.